File: src/main/java/com/github/dwendelen/platformd/core/budget/IncomeSource.java
package com.github.dwendelen.platformd.core.budget;

import com.github.dwendelen.platformd.core.budget.command.CreateExpensePost;
import com.github.dwendelen.platformd.core.budget.command.CreateIncomeSource;
import com.github.dwendelen.platformd.core.budget.command.RecordIncome;
import com.github.dwendelen.platformd.core.budget.event.ExpensePostCreated;
import com.github.dwendelen.platformd.core.budget.event.IncomeRecorded;
import com.github.dwendelen.platformd.core.budget.event.IncomeSourceCreated;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.commandhandling.model.AggregateIdentifier;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.spring.stereotype.Aggregate;

import java.util.UUID;

import static org.axonframework.commandhandling.model.AggregateLifecycle.apply;

@Aggregate
public class IncomeSource {
    @AggregateIdentifier
    private UUID uuid;

    public IncomeSource() {}

    @CommandHandler
    public IncomeSource(CreateIncomeSource command) {
        apply(new IncomeSourceCreated()
                .setUuid(command.getUuid())
                .setName(command.getName())
        );
    }

    @EventSourcingHandler
    public void on(IncomeSourceCreated event) {
        this.uuid = event.getUuid();
    }

    @CommandHandler
    public void handle(RecordIncome command) {
        apply(new IncomeRecorded()
                .setIncomeSource(command.getIncomeSource())
                .setAmount(command.getAmount())
                .setAccountId(command.getAccountId())
                .setTransactionId(command.getTransactionId())
        );
    }

    @EventSourcingHandler
    public void on(IncomeRecorded event) {

    }
}

File: src/main/scala/com/github/dwendelen/platformd/core/budget/IncomeSource.scala
package com.github.dwendelen.platformd.core.budget

import java.util.UUID

import com.fasterxml.jackson.annotation.JsonCreator
import com.github.dwendelen.platformd.infrastructure.cqrs.{Aggregate, AggregateResult, CommandSucces}

class IncomeSource(uuid: UUID) extends Aggregate[IncomeSource] {
    def this() = this(null)

    override def applyEvent(event: Any): IncomeSource = event match {
        case e: IncomeSourceCreated => applyEvent(e)
        case _ => this
    }

    def applyEvent(event: IncomeSourceCreated): IncomeSource =
        new IncomeSource(event.uuid)

    override def handle(command: Any): AggregateResult = command match {
        case cmd: CreateIncomeSource => AggregateResult(CommandSucces(cmd.uuid),handle(cmd))
        case cmd: RecordIncome => AggregateResult(CommandSucces(uuid), handle(cmd))
    }
    def handle(cmd: CreateIncomeSource): List[Any] =
        List(new IncomeSourceCreated(
            cmd.uuid,
            cmd.owner,
            cmd.name
        ))

    def handle(cmd: RecordIncome): List[Any] =
        List(new IncomeRecorded(
            cmd.incomeSource,
            cmd.transactionId,
            cmd.amount,
            cmd.accountId
        ))
}

//Events
@JsonCreator
class IncomeSourceCreated(val uuid: UUID,
                          val owner: UUID,
                          val name: String)
@JsonCreator
class IncomeRecorded(val incomeSource: UUID,
                     val transactionId: UUID,
                     val amount: BigDecimal,
                     val accountId: UUID)

//Commands
@JsonCreator
class CreateIncomeSource(val  uuid: UUID,
                         val owner: UUID,
                         val name: String)
@JsonCreator
class RecordIncome(val incomeSource: UUID,
                   val transactionId: UUID,
                   val amount: BigDecimal,
                   val accountId: UUID)